### PR TITLE
feat(dashboardgrid): allow resizing by dragging the border

### LINKED ISSUE #1598: [DashboardGrid / Card] allow resizing by dragging the border
### What package is this for?

- [x] React
- [ ] Angular

### Summary

Currently, the only way to change the size is to do it programmatically. It could be beneficial to allow the user to drag the border of a card while in a DashboardGrid to resize it and see the changes or previews live.

This would probably only be worth effort if we can truly lock the resizing to each of our predefined sizes. If we have to actually try to adapt our styling to each completely custom size, it would be costly

**Additional context**
React grid layout allows resizing of items and allows min and max widths / heights so I believe this is possible.

Related props: https://github.com/STRML/react-grid-layout#grid-item-props

Related demo: https://strml.github.io/react-grid-layout/examples/0-showcase.html

### Specific timeline issues / requests

Do you want this work within a specific time period? Is it related to an
upcoming release?

No specific timeline

### Available extra resources

What resources do you have to assist this effort?

_Carbon add-ons for Watson IoT is a collaborative system. We encourage teams to build components and submit them for integration._

### PR DESCRIPTION
**Pull request by bjornalm**
Closes #1598 

**Summary**
Added resize functionality to all cards when present in a DashboardGrid. The feature is based on the underlying resize functionality of the react-grid-layout which passes down a resize handler to the grid items (cards). The DashboardGrid listens to all resize callbacks and makes sure the only valid card sizes are suggested and available as drop while dragging.

The actual card content is showing during resize and I believe performance is acceptable. In the stories added the card also gets transformed between the defined card sizes during the resize drag process.

**Change List (commits, features, bugs, etc)**
- Modified DashboardGrid to manage card resizing using strict card sizes. It now exposes onCardSizeChange and onResizeStop. Both callbacks include the current name and size of the card + all additional parameters passed from the react-grid-layout that might be needed to manage/store the new state of the grid. 
- Modified Card to wrap the onStart/onStop callbacks of the resize handle so that the card can know when it is resized (and show resize border)
- Modified all the card components to extract the resize handles from children and explicitly pass them via a special prop to the underlying Card component. This was needed since most cards don't pass down the children to the Card and some pass children as a function. 
- All cards (including the base Card) has a isResizable prop that controls if the card is resizable in the DashboardGrid.
- Added useMemo to PieChart so that the sample data isn't regenerated during resize
- Removed the mock in the mock of componentUtilityFunctions.js in the DashboardGrid.test and tested the result instead.
- Created unit tests
- Created a story that shows a single card that can be resized. 
- Created a story that shows all the card in different sizes where some cards have minimum sizes defined.
- adjusted the CSS to be more inline with the PAL design
- adjusted the coverage threshold 

**Acceptance Test (how to verify the PR)**
I tested in Chrome, Safari & Firefox.
Checkout out the two stories and plat around with the resizing and knobs. Verify that performance and functionality is ok.

https://deploy-preview-1722--carbon-addons-iot-react.netlify.app?path=/story/watson-iot-dashboard-grid--dashboard-all-cards-as-resizable

https://deploy-preview-1722--carbon-addons-iot-react.netlify.app?path=/story/watson-iot-dashboard-grid--dashboard-resizable-card

**Note!**
If we upgrade to the latest version (1.1.1) we can use the horizontal (west) & vertical (south) border resize as shown in the PAL design. I've tried it and it works, but since it requires a major upgrade of the react-grid-layout I will create a separate issue for that.

### CODE REVIEW

**Review by scottdickerson — APPROVED** [review]
LGTM

File: jest.config.js [hunk 1]
```diff
@@ -13,12 +13,17 @@ module.exports = {
...
       lines: 80,
     },
-    './src/components/**/!(ColumnHeaderSelect|FilterHeaderRow|TableToolbar|RowActionsCell|RowActionsError|StatefulTable|StatefulTableDetailWizard|CatalogContent|FileDrop|HeaderMenu|Dashboard|CardRenderer|Attribute|UnitRenderer|ImageHotspots|ImageControls|PageHero|PageTitle|EditPage|AsyncTable|ImageCard|WizardHeader|TableHead|ColumnResize|DateTimePicker|TimeSeriesCard|BarChartCard).jsx': {
+    './src/components/**/!(ColumnHeaderSelect|FilterHeaderRow|TableToolbar|RowActionsCell|RowActionsError|StatefulTable|StatefulTableDetailWizard|CatalogContent|FileDrop|HeaderMenu|Dashboard|CardRenderer|Attribute|UnitRenderer|ImageHotspots|ImageControls|PageHero|PageTitle|EditPage|AsyncTable|ImageCard|WizardHeader|TableHead|ColumnResize|DateTimePicker|TimeSeriesCard|BarChartCard|DashboardGrid).jsx': {
       statements: 80,
       branches: 80,
       functions: 80,
       lines: 80,
     },
+    './src/components/Dashboard/DashboardGrid.jsx': {
```
Comment (davidicus) [inline jest.config.js:22]: Why did we have to make exceptions here?

Comment (bjornalm) [reply jest.config.js:22]: I found it difficult to write tests with React Testing Library and their strategy (test the way a user would test) in combination with the underlying react-grid-layout library. I was not able to get drag events working in the test environments and exposing additional functions by export required refactoring that made the code worse.

I can a) spend some more time on it to se if I can find a way to proper test it, b) use enzyme, c) leave as is

Comment (tay1orjones) [reply jest.config.js:22]: Yeah that makes sense - I think it would be better tested via cypress in a real browser #1203

**Review by tay1orjones — COMMENTED** [review]
I'm not sure if there's any way to improve this, but at times the reflow on resize is a bit jarring and not what I'd expect.

<details>
  <summary>Here I'm just making it smaller, but when I let go it reflows all wonky.</summary>

  auto reflow
</details>






<details>
  <summary>Other times if I get it on a boundary it causes some wild reflow render flashing</summary>

wonky
</details>

**Comment by bjornalm** [comment]
> I'm not sure if there's any way to improve this, but at times the reflow on resize is a bit jarring and not what I'd expect.
> 
> Here I'm just making it smaller, but when I let go it reflows all wonky.
> auto reflow
> 
For this story I haven't implemented a state handler for the reordering of cards, only for the resize, so I guess that you had reordered your cards and when you then resize a card the new card size is loaded with the layout of the original positions. The easiest fix is probably that I disable the 'isEditable' option for this story, alternately implement the state of the reordering as well.  

> Other times if I get it on a boundary it causes some wild reflow render flashing
I might be able to mitigate or fix this. I will investigate.

**Comment by tay1orjones** [comment]
@bjornalm 👍  If there's a way for the consumer to provide a state handler for the reordering of cards, we can defer those pieces into separate issues if you prefer

File: src/components/Dashboard/DashboardGrid.jsx [hunk 2]
```diff
@@ -141,6 +221,8 @@ const DashboardGrid = ({
   layouts,
   onLayoutChange,
   onBreakpointChange,
+  onCardSizeChange,
+  onResizeStop: onResizeStopCallback, // TODO: remove?
```
Comment (tay1orjones) [inline src/components/Dashboard/DashboardGrid.jsx:225]:
```suggestion
  onResizeStop: onResizeStopCallback,
```

**Review by tay1orjones — APPROVED** [review]

**Comment by bjornalm** [comment]
> @bjornalm 👍 If there's a way for the consumer to provide a state handler for the reordering of cards, we can defer those pieces into separate issues if you prefer

We never created such a story, but I assume that clients that use the `isEditable` of the DashboardGrid has implemented a state handler so that they can actually save the new layout, otherwise it wouldn't be very useful. Should I create an issue for adding such a story **and** story for `isEditable` combined with this new resize functionality?

**Comment by bjornalm** [comment]
@tay1orjones I think I managed to get the crazy reflow fixed. We now calculate for the possibility that a card is jumping around during resize and that the x and y value will change as a consequence. The fix compensates for that so that for example a card that is pushed to a new row because it grew big will have the dropzone stay that new size. You can check it out here
https://deploy-preview-1722--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-dashboard-grid--dashboard-all-cards-as-resizable

**Comment by tay1orjones** [comment]
@bjornalm Wow, yes that is such an improvement!

And yes you're right about `isEditable`, sorry for my confusion. I don't think we need a story showcasing how to wrap a prop in state to toggle on/off. A story for isEditable combined with this new resize functionality could be useful, though not a requirement for this PR.

**Pull request merged**